docs/design/2021-06-23-restricted-read-only.md

```diff
@@ -0,0 +1,48 @@
+# Technical Design Document for Cloud Read Only
+
+> Author: [huzhifeng@pingcap.com](mailto:huzhifeng@pingcap.com)
+This document is open to the web, and it will appear as a design doc PR to the TiDB repository.
+
+# Introduction
+
+To allow Cross-Region Replication on TiDB Cloud, we need a more strict level of read-only. In this document, we introduced the `TIDB_RESTRICTED_READ_ONLY` global variable. Turning it on will make the cluster read-only eventually for all users, including users with SUPER or CONNECTION_ADMIN privilege.
+
+# Motivation or Background
+
+See the [original specification](https://docs.google.com/document/d/1LtmW4fIohTjPcupD9smQGQpN2B2tRSFNyH8PFbg-L2c/edit).
+
+# **Detailed Design**
+
+## read-only handling
+
+Currently, TiDB supports `read_only` and `super_read_only` global variables with no actual operation. Since TIDB_RESTRICTED_READ_ONLY does not comply with the behavior of MySQL read-only and super-read-only (which will block if there is table locked or ongoing committing of a transaction), we will not build TIDB_RESTRICTED_READ_ONLY upon read-only and super-read-only. These two variables will be remained untouched.
+
+We will create a new global variable TIDB_RESTRICTED_READ_ONLY, and turning it to on and off will only be allowed in SEM mode, and only users with RESTRICTED_VARIABLES_ADMIN will be allowed to modify this variable.
+
+To allow replication service to write, we introduced a new dynamic privilege level called `RESTRICTED_REPLICA_WRITER_ADMIN`, user with this privilege will be ignored for read-only checks.
+
+All TiDB servers in the cluster will periodically update their own cache of the global variable table, so if we change the global variable, eventually all TiDB servers will see the change.
```

Comment:
This is quite deterministic, so I think we can mention how quickly it will take effect (it is instant on the server you are connected to, and starts refresh of the system variables asyncronously on all other servers my etcd notification. In the event that a server misses an etcd event, it may take up to 30s).

See: https://github.com/pingcap/tidb/pull/24359

Another small clarification:

Because the sysvar is *only* global scoped and not session scoped, it will also take effect for existing sessions. TiDB previously cached global-only scoped variables in the session, but no longer does https://github.com/pingcap/tidb/issues/24368 ; this is essential for working correctly with connection pooled environments (both connection pooled and non-pooled scenario tests will be required in testing.)